tools/harness-automation/autothreadharness/harness_case.py

```diff
@@ -476,10 +480,56 @@ def _test_bed(self):
         devices = [device for device in settings.GOLDEN_DEVICES
                    if not self.history.is_bad_golden_device(device[0]) and \
                    not (settings.DUT_DEVICE and device[0] == settings.DUT_DEVICE[0])]
-
         logger.info('Available golden devices: %s', json.dumps(devices, indent=2))
         golden_devices_required = self.golden_devices_required
 
+        #for mixed test bed
+        if settings.GOLDEN_DEVICE_TYPE and re.match('mixed', settings.GOLDEN_DEVICE_TYPE, re.M|re.I):
+            topo_file = settings.HARNESS_HOME+"\\Thread_Harness\\TestScripts\\TopologyConfig.txt"
+            try:
+                f_topo = open(topo_file,'r')
+            except IOError,e:
+                logger.info('%s can NOT be found',topo_file)
+                raise GoldenDeviceNotEnoughError()
+            topo_mixed_devices=[]
+            try:
+                while 1:
+                    topo_line = f_topo.readline().strip()
+                    logger.info('%s',topo_line)
+                    match_line = re.match(r'(.*)-(.*)', topo_line, re.M | re.I)
+                    case_id = match_line.group(1)
+
+                    if re.sub(r'\.',' ',case_id) == self.case:
+                        topo_device_list = re.split(',', match_line.group(2))
+                        for i in range(len(topo_device_list)):
+                            topo_device = re.split(':', topo_device_list[i])
+                            topo_mixed_devices.append(tuple(topo_device))
+                        break
+                    else:
+                        continue
+            except:
+                logger.info('Get devices from topology config file error')
+                raise GoldenDeviceNotEnoughError()
+            logger.info('Topology config devices for case %s: %s',case_id, topo_mixed_devices)
+            f_topo.close()
+            needed_golden_devices = []
+            # mapping topology config devices with devices in settings
+            for temp_mixed_device in topo_mixed_devices:
+                for temp_device in devices:
```

Comment:
Any possibility of shortage for device from some vendor?

Reply:
I think it would be more helpful for our automation tool to prompt how many and which kind of devices are in absence, relieving the user from checking the topology configuration file himself, specially for some large scale cases.

Reply:
It is fine to file a separate PR to improve this.